pkg/globalnet/controllers/ipam/ipam.go

```diff
@@ -0,0 +1,308 @@
+
+	// Wait for the caches to be synced before starting workers
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, i.servicesSynced, i.podsSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	klog.Info("Starting workers")
+	go wait.Until(i.runServiceWorker, time.Second, stopCh)
+	go wait.Until(i.runPodWorker, time.Second, stopCh)
+	wg.Wait()
+	<-stopCh
+	klog.Info("Shutting down workers")
+
+	return nil
+}
+
+func (i *Controller) runServiceWorker() {
+	for i.processNextObject(i.serviceWorkqueue, i.serviceGetter, i.serviceUpdater) {
+	}
+}
+
+func (i *Controller) runPodWorker() {
+	for i.processNextObject(i.podWorkqueue, i.podGetter, i.podUpdater) {
+	}
+}
+
+func (i *Controller) processNextObject(objWorkqueue workqueue.RateLimitingInterface, objGetter func(namespace, name string) (runtime.Object, error), objUpdater func(runtimeObj runtime.Object, key string) error) bool {
+	obj, shutdown := objWorkqueue.Get()
+	if shutdown {
+		return false
+	}
+	err := func() error {
+		defer objWorkqueue.Done(obj)
+
+		key := obj.(string)
+		ns, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			objWorkqueue.Forget(obj)
+			return fmt.Errorf("error while splitting meta namespace key %s: %v", key, err)
+		}
+
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// Retrieve the latest version of object before attempting update
+			// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+			runtimeObj, err := objGetter(ns, name)
+
+			if err != nil {
+				// mostly this means object already deleted
+				objWorkqueue.Forget(obj)
+				return fmt.Errorf("error retrieving submariner-ipam-controller object %s: %v", name, err)
+			}
+			return objUpdater(runtimeObj, key)
+		})
+		if retryErr != nil {
+			// failed to get globalIp or failed to update, we want to retry
+			return retryErr
+		}
+
+		objWorkqueue.Forget(obj)
+		return nil
+	}()
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+func (i *Controller) enqueueObject(obj interface{}, workqueue workqueue.RateLimitingInterface) {
+	if key := i.allowEnqueue(obj); key != "" {
+		klog.V(4).Infof("Enqueueing %v for ipam controller", key)
+		workqueue.AddRateLimited(key)
+	}
+}
+
+func (i *Controller) allowEnqueue(obj interface{}) string {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return ""
+	}
+	ns, _, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return ""
+	}
+	if i.excludeNamespaces[ns] {
+		return ""
+	}
+	return key
+}
+
+func (i *Controller) handleUpdateService(old interface{}, new interface{}) {
+	service := new.(*k8sv1.Service)
+	oldGlobalIp := old.(*k8sv1.Service).GetAnnotations()[submarinerIpamGlobalIp]
+	newGlobalIp := new.(*k8sv1.Service).GetAnnotations()[submarinerIpamGlobalIp]
```

Comment:
`new.(*k8sv1.Service)` you already have it in `service`

Reply:
Argh! Looks like one of my changes got lost when rebasing. this is supposed to be key and not service name anymore.